### ISSUE TITLE
With an external grouper, there is no way to access the grouped value in a DataFrame(...).groupby(...).apply(...) workflow

### ISSUE THREAD

**Issue by brianthelion** [pydata/pandas#9545]
groupby-apply workflows are important pandas idioms. Here's a brief example grouping on a named DataFrame column:
```
>>> df = pd.DataFrame({'key': [1, 1, 1, 2, 2, 2, 3, 3, 3], 'value': range(9)})
>>> result = df.groupby('key').apply(lambda x: x['key'])
>>> result
key   
1    0    1
     1    1
     2    1
2    3    2
     4    2
     5    2
3    6    3
     7    3
     8    3
Name: key, dtype: int64
```
An important highlight of this example is the ability to reference the grouped value -- eg, `x['key']` -- inside the applied function.

pandas also supports grouping on arbitrary mapping functions, iterables, and lots of other objects. In these cases, the grouped value is not represented as a named column in the DataFrame. Thus, when using apply(...), there is no apparent way to access the group key value. The only alternative is to use a (slow) for-loop as in:
```
foo = lambda _k, _g: ...
grouped = df.groupby(grouper)
result_iter = (foo(key, group) for key, group in grouped) 
key_iter = (key for key, group in df.groupby(grouper) in grouped)
pd.DataFrame.from_records(result_iter, index=key_iter)
```
IMHO, the ability to access the grouped value in an idiomatic way from with applied function is ergonomically important; the groupby-apply idiom is at best partially realized without it.

**Comment by jreback** [comment]
see #9239 which implements the copy-assign idiom

**Comment by brianthelion** [comment]
@jreback Sorry being a complete n00b here, but at a quick glance I can't imagine how to solve the present issue with copy-assign. My imagination is pretty limited, though. Could you provide a basic intuition? Thanks!

**Comment by TomAugspurger** [comment]
Do you have a small example where you need to reference the grouping key? I'm trying to see where this would be necessary (or just convenient).

Also, does the `.name` attribute on the values of the groupby help you out?

**Comment by jreback** [comment]
```
In [9]: df.groupby('key').apply(lambda x: Series(x.name,x.index))
Out[9]: 
key   
1    0    1
     1    1
     2    1
2    3    2
     4    2
     5    2
3    6    3
     7    3
     8    3
dtype: int64
```

**Comment by brianthelion** [comment]
It appears that .name is the attribute that I've been looking for! Is this the intended use-case for the attribute? If so, my feeling is that "name" is somewhat nondescript.

**Comment by TomAugspurger** [comment]
It is the intended use. I think it's documented in the transform section.

**Comment by TomAugspurger** [comment]
I just looked through the docs and didn't see anything about the `.name` being set. It *is* in the docstring for `gr.transform`

> Each subframe is endowed with the attribute 'name' in case you need to know which group you are working on

@brianthelion are you interested in submitting a pull-request to clarify the prose docs (and maybe the docstrings on apply?) I don't think the `.name` attribute is set for `.agg` operations. Not sure if this is intentional.

**Comment by brianthelion** [comment]
Happy to! However, I think there should first be some discussion about the semantics of the attribute. IMHO, "name" is way too generic and semantically kinda nonsensical in lots of use-cases. I'm not sure what's appropriate for the others, but ".grouped_value" or thereabouts seems appropriate for the groupby-apply workflow.

**Comment by TomAugspurger** [comment]
I'd vote for `.key`.* `*.name` was probably chosen because it already has a
meaning on Series.

Is there any risk of clobbering an attribute on someone's Series? I think
we're ok, unless they're explicitly setting something to `.key` in the
apply function...

**Comment by alkasm** [comment]
> Is there any risk of clobbering an attribute on someone's Series?

@TomAugspurger yes :( Just happened to me. Not sure how I feel about an attribute with such a common name added to the dataframe which wasn't there in the ungrouped dataframe. Was a really hard bug to pinpoint.

https://github.com/pandas-dev/pandas/issues/25457

**Comment by ericabrauer** [comment]
> @ericabrauer FYI, probably better to add that to the issue I linked, which is more aligned with identifying the `.name` attribute as an issue.

Oh ok- thanks @alkasm, I guess maybe I don't have a full enough grasp on the issue. I'll try to take a look at the one you referenced and see if I can better understand.

**Comment by Vishnu-sai-teja** [comment]
I am new to Open Source Contribution , but have a good knowledge of how pandas work , I just understand the issue compared to that of other , I just want to try .

**Comment by Vishnu-sai-teja** [comment]
take

**Comment by samyarpotlapalli** [comment]
take

**Comment by 272006Sakshi** [comment]
> groupby-apply workflows are important pandas idioms. Here's a brief example grouping on a named DataFrame column:
> 
> ```
> >>> df = pd.DataFrame({'key': [1, 1, 1, 2, 2, 2, 3, 3, 3], 'value': range(9)})
> >>> result = df.groupby('key').apply(lambda x: x['key'])
> >>> result
> key   
> 1    0    1
>      1    1
>      2    1
> 2    3    2
>      4    2
>      5    2
> 3    6    3
>      7    3
>      8    3
> Name: key, dtype: int64
> ```
> 
> An important highlight of this example is the ability to reference the grouped value -- eg, `x['key']` -- inside the applied function.
> 
> pandas also supports grouping on arbitrary mapping functions, iterables, and lots of other objects. In these cases, the grouped value is not represented as a named column in the DataFrame. Thus, when using apply(...), there is no apparent way to access the group key value. The only alternative is to use a (slow) for-loop solution as in:
> 
> ```
> foo = lambda _k, _g: ...
> grouped = df.groupby(grouper)
> result_iter = (foo(key, group) for key, group in grouped) 
> key_iter = (key for key, group in grouped)
> pd.DataFrame.from_records(result_iter, index=key_iter)
> ```
> 
> IMHO, the ability to access the grouped value in an idiomatic way from within the applied function is ergonomically important; the groupby-apply idiom is at best partially realized without it.

Hi, I’m Sakshi, new to Pandas contributions.
From reading the discussion, my understanding is that this issue is mainly about improving documentation around the `.name` attribute being available inside `GroupBy.apply`, since this behavior isn’t very discoverable in the current docs.

I’d like to work on clarifying the documentation by adding a short explanation, a simple example, and a note about possible ambiguity when a column named `"name"` exists. Please let me know if this sounds reasonable.

**Comment by Lk-ux** [comment]
take
